compiler/generator_rust.c

```diff
@@ -0,0 +1,1347 @@
+}
+
+static void generate_repeat(struct generator * g, struct node * p, struct str * loopvar) {
+
+     struct str * savevar = vars_newname(g);
+    int keep_c = repeat_restore(g, p->left);
+    int replab = new_label(g);
+    g->I[0] = replab;
+    write_comment(g, p);
+    writef(g, "~M'replab~I0: loop{~N~+", p);
+
+    if (keep_c) write_savecursor(g, p, savevar);
+
+    g->failure_label = new_label(g);
+    str_clear(g->failure_str);
+    g->I[0] = g->failure_label;
+    w(g, "~M'lab~I0: for _ in 0..1 {~N~+");
+    generate(g, p->left);
+
+    if (!g->unreachable) {
+        if (loopvar != 0) {
+            g->B[0] = str_data(loopvar);
+            w(g, "~M~B0 -= 1;~N");
+        }
+
+        g->I[0] = replab;
+        w(g, "~Mcontinue 'replab~I0;~N");
+    }
+    w(g, "~-~M}~N");
+    g->unreachable = false;
+
+    if (keep_c) write_restorecursor(g, p, savevar);
+
+    g->I[0] = replab;
+    w(g, "~Mbreak 'replab~I0;~N~-~M}~N");
+    str_delete(savevar);
+}
+
+static void generate_atleast(struct generator * g, struct node * p) {
+
+    struct str * loopvar = vars_newname(g);
+    write_comment(g, p);
+    g->B[0] = str_data(loopvar);
+    w(g, "~Mlet mut ~B0 = ");
+    generate_AE(g, p->AE);
+    w(g, ";~N");
+    {
+        int a0 = g->failure_label;
+        struct str * a1 = str_copy(g->failure_str);
+
+        generate_repeat(g, p, loopvar);
+
+        g->failure_label = a0;
+        str_delete(g->failure_str);
+        g->failure_str = a1;
+    }
+    g->B[0] = str_data(loopvar);
+    write_failure_if(g, "~B0 > 0", p);
+    str_delete(loopvar);
+}
+
+static void generate_setmark(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->V[0] = p->name;
+    writef(g, "~M~V0 = env.cursor;~N", p);
+}
+
+static void generate_tomark(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? ">" : "<";
+
+    w(g, "~Mif env.cursor ~S0 "); generate_AE(g, p->AE);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+    w(g, "~Menv.cursor = "); generate_AE(g, p->AE); writef(g, ";~N", p);
+}
+
+static void generate_atmark(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    w(g, "~Mif env.cursor != "); generate_AE(g, p->AE);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+}
+
+
+static void generate_hop(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? "" : "-";
+
+    w(g, "~Mlet c = env.byte_index_for_hop(~S0");
+    generate_AE(g, p->AE);
+    w(g, ");~N");
+
+    g->S[0] = p->mode == m_forward ? "0" : "env.limit_backward";
+
+    write_failure_if(g, "~S0 as i32 > c || c > env.limit as i32", p);
+    writef(g, "~Menv.cursor = c as usize;~N", p);
+}
+
+static void generate_delete(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    writef(g, "~Mif !env.slice_del(){~N"
```

Comment:
It would be better style to change `){` to `) {`. The same is true in `generate_slicefrom`.